src/main/java/com/artipie/asto/SubStorage.java

```diff
@@ -53,8 +53,14 @@ public SubStorage(final Key prefix, final Storage origin) {
 
     @Override
     public CompletableFuture<Collection<Key>> list(final Key filter) {
+        final Key cleanfilter;
+        if (this.prefix == Key.ROOT && filter == Key.ROOT) {
+            cleanfilter = Key.ROOT;
+        } else {
+            cleanfilter = new PrefixedKed(this.prefix, filter);
+        }
```

Comment:
What is a reason for these changes?

Reply:
It's to avoid exception :
```cmd
io.grpc.StatusRuntimeException: INVALID_ARGUMENT: etcdserver: key is not provided
```
while trying to list `ROOT` key of an Etcd's `SubStorage` with `ROOT` as prefix.
We got this exception because `new PrefixedKed(Key.ROOT, Key.ROOT)` is not equal to `Key.ROOT` as expected here (in `EtcdStorage`) :
```java
@Override
public CompletableFuture<Collection<Key>> list(final Key prefix) {
    final CompletableFuture<GetResponse> future;
    if (prefix == Key.ROOT) {
        future = this.client.getKVClient().get(
            EtcdStorage.ETCD_ROOT_KEY,
            GetOption.newBuilder()
                .withKeysOnly(true)
                .withSortOrder(SortOrder.ASCEND)
                .withRange(EtcdStorage.ETCD_ROOT_KEY)
                .build()
        );
    } else {
        future = this.client.getKVClient().get(
```
Perhaps, it would be better to do like this :
```java
if (prefix.equals(Key.ROOT))
```
and implements method `equals` of `Key`.

Reply:
@baudoliver7 yes, I think it's a good direction to make keys equitable & comparable by its values, maybe event better to use `interface Key extends Comparable<Key>` and use its `parts()` method to compare

Reply:
Ok, but while looking at `Key.From`, I noticed that methods `equals` and `hashCode` have already been implemented :
```java
@Override
public boolean equals(final Object another) {
    if (this == another) {
        return true;
    }
    if (another == null || getClass() != another.getClass()) {
        return false;
    }
    final From from = (From) another;
    return Objects.equals(this.parts, from.parts);
}

@Override
public int hashCode() {
    return Objects.hash(this.parts);
}
```
but decorated classes don't use them because `Key.Wrap` doesn't override them.

I will override them and make `Key` comparable also. Are you okay with ?

Reply:
@g4s8 Done in my last changes.